module-system/module-implementations/sov-evm/src/query.rs

```diff
@@ -25,42 +27,93 @@ impl<C: sov_modules_api::Context> Evm<C> {
         Ok(Some(reth_primitives::U64::from(1u64)))
     }
 
-    // TODO https://github.com/Sovereign-Labs/sovereign-sdk/issues/502
     #[rpc_method(name = "getBlockByNumber")]
     pub fn get_block_by_number(
         &self,
-        _block_number: Option<String>,
-        _details: Option<bool>,
-        _working_set: &mut WorkingSet<C::Storage>,
+        block_number: Option<String>,
+        details: Option<bool>,
+        working_set: &mut WorkingSet<C::Storage>,
     ) -> RpcResult<Option<reth_rpc_types::RichBlock>> {
         info!("evm module: eth_getBlockByNumber");
 
-        let header = reth_rpc_types::Header {
-            hash: Default::default(),
-            parent_hash: Default::default(),
-            uncles_hash: Default::default(),
-            miner: Default::default(),
-            state_root: Default::default(),
-            transactions_root: Default::default(),
-            receipts_root: Default::default(),
-            logs_bloom: Default::default(),
-            difficulty: Default::default(),
-            number: Default::default(),
-            gas_limit: Default::default(),
-            gas_used: Default::default(),
-            timestamp: Default::default(),
-            extra_data: Default::default(),
-            mix_hash: Default::default(),
-            nonce: Default::default(),
-            base_fee_per_gas: Some(reth_primitives::U256::from(100)),
-            withdrawals_root: Default::default(),
+        // "safe" and "finalized" are not implemented
+        let block = match block_number {
+            Some(ref block_number) if block_number == "earliest" => self
+                .blocks
+                .get(0, &mut working_set.accessory_state())
+                .expect("Genesis block must be set"),
+            Some(ref block_number) if block_number == "latest" => self
+                .head
+                .get(working_set)
+                .expect("Head block must be set")
+                .seal(),
+            Some(ref block_number) if block_number == "pending" => self
+                .pending_head
+                .get(&mut working_set.accessory_state())
+                .expect("Pending head block must be set")
+                .seal(),
```

Comment:
I think we won't support `pending` as it doesn't make sense for us, for `pending` block we would have to get the next DA block.

Reply:
I am not sure about that. There will be some rollups that use single centralized sequencer to give soft confirmations, such as ours. In that case, there will be a pending block on the sequencer level which is not broadcasted or sent to the DA layer.

Reply:
ok but then it should live on the sequencer and not on the module. Sequencer can extend the api, it will be a separate component that can us the module.. I don't see this doable inside the module.

Reply:
I think we need to merge [#870](https://github.com/Sovereign-Labs/sovereign-sdk/pull/870) before writing a test. [@orkunkilic](https://github.com/orkunkilic) let's discuss the use case you have on our sync-call. I am sure we will accommodate it. In the meantime, I'm ok with two options: we can either merge this PR now and make a comment on #878 that we need to further work on `test` and the `pending block` case, or we can wait to discuss it during our upcoming sync-up call, whichever you prefer, [@orkunkilic](https://github.com/orkunkilic). Thank you for your excellent contribution!

Reply:
I think you can write test without it, I written receipts and transactions tests without it. In full node it should work.